### PR TITLE
Fix error when a job exits with something other than a stacktrace

### PR DESCRIPTION
**Pull request by derekkraan**
We have seen that sometimes Oban is unable to process certain `DOWN` messages. It crashes on a line calling `Exception.format` because the "stacktrace" that we save here is not always a stacktrace after all (see the added test for example).

From https://hexdocs.pm/elixir/Exception.html#format/3, if `kind` is `{:EXIT, pid}`, then it doesn't format the stacktrace.

I'm reasonably certain that this is the correct behaviour but curious what your thoughts are.

### CODE REVIEW

**Comment by derekkraan** [comment]
CI is failing on cyclomatic complexity being greater than 10 of a function in `test/support/worker.ex`. I can refactor this function but disabling this check for this function is also an option, what do you think @sorentwo ?

File: test/support/worker.ex [hunk 1]
```diff
@@ -61,6 +61,13 @@ defmodule Oban.Integration.Worker do
         |> Task.await()
 
         :ok
+
+      "TASK_EXIT_ERROR" ->
```
Comment (MvanDiemen) [inline test/support/worker.ex:65]: Where does this one come from?

**Comment by MvanDiemen** [comment]
Good find that it doesn't generate a stacktrace

File: lib/oban/queue/producer.ex [hunk 2]
```diff
@@ -100,7 +100,7 @@ defmodule Oban.Queue.Producer do
     # block the producer while awaiting a retry.
     Task.Supervisor.async_nolink(state.foreman, fn ->
       Breaker.with_retry(fn ->
-        %{exec | kind: :error, error: error, stacktrace: stack, state: :failure}
+        %{exec | kind: {:EXIT, pid}, error: error, stacktrace: stack, state: :failure}
```
Comment (sorentwo) [inline lib/oban/queue/producer.ex:103]: The `kind` is exposed through telemetry data, and meant to match the values that come from `telemetry.span`. What if we keep this as `:error` but ensure that `stack` is an empty list?

**Comment by sorentwo** [comment]
> CI is failing on cyclomatic complexity being greater than 10 of a function in `test/support/worker.ex`. I can refactor this function but disabling this check for this function is also an option, what do you think @sorentwo ?

👍 I'd rather disable the check (especially for a test module).

File: lib/oban/queue/producer.ex (hunk 2 above)
Comment (derekkraan) [reply lib/oban/queue/producer.ex:103]: Options:
1. kind: `:error`: `"error" => "** (ErlangError) Erlang error: :timeout"`
2. kind: `:exit` (also a valid `kind` according to `telemetry` docs): `"error" => "** (exit) time out"`
3. kind: `{:EXIT, pid}`: `"error" => "** (EXIT from #PID<0.3249.0>) time out" `

I'm not quite sure what happens in Telemetry when you pass `{:EXIT, pid}`.

Comment (derekkraan) [reply lib/oban/queue/producer.ex:103]: I looked into this further, `Exception.kind` has type `:error | :exit | :throw | {:EXIT, pid}`, whilst telemetry kind only includes `:error | :exit | :throw`. I think it's a bit weird. We are using `kind` both in calls to the `Exception` module and in calls to telemetry.

I added a little case statement to convert from `Exception.kind` to the `metadata.kind` that `:telemetry` is expecting. This gives us the best error message, and still conforms to telemetry's spec.

Let me know if you'd still prefer `:exit`.

File: lib/oban/queue/executor.ex [hunk 3]
```diff
@@ -170,10 +170,16 @@ defmodule Oban.Queue.Executor do
   def emit_event(%__MODULE__{state: :failure} = exec) do
     measurements = %{duration: exec.duration, queue_time: exec.queue_time}
 
+    kind =
+      case exec.kind do
```
Comment (sorentwo) [inline lib/oban/queue/executor.ex:174]: Initially, I wasn't sure about this. Now that I see it, I appreciate the explicitness. Nice work 👍

File: test/integration/executing_test.exs [hunk 4]
```diff
@@ -71,7 +71,8 @@ defmodule Oban.Integration.ExecutingTest do
 
   defp job do
     gen all queue <- member_of(~w(alpha beta gamma delta)),
-            action <- member_of(~w(OK DISCARD ERROR EXIT FAIL KILL SNOOZE TASK_ERROR)),
+            action <-
+              member_of(~w(OK DISCARD ERROR EXIT FAIL KILL SNOOZE TASK_ERROR TASK_EXIT_ERROR)),
```
Comment (sorentwo) [inline test/integration/executing_test.exs:75]: The other action is `EXIT`, not `EXIT_ERROR` (and this keeps it on a single line still).

```suggestion
            action <- member_of(~w(OK DISCARD ERROR EXIT FAIL KILL SNOOZE TASK_ERROR TASK_EXIT),
```

File: test/support/worker.ex [hunk 5]
```diff
@@ -61,6 +62,13 @@ defmodule Oban.Integration.Worker do
...
```
Comment (sorentwo) [inline test/support/worker.ex:66]:
```suggestion
      "TASK_EXIT" ->
```

File: test/integration/executing_test.exs [hunk 6]
```diff
@@ -71,7 +71,7 @@ defmodule Oban.Integration.ExecutingTest do
 
   defp job do
     gen all queue <- member_of(~w(alpha beta gamma delta)),
-            action <- member_of(~w(OK DISCARD ERROR EXIT FAIL KILL SNOOZE TASK_ERROR)),
+            action <- member_of(~w(OK DISCARD ERROR EXIT FAIL KILL SNOOZE TASK_ERROR TASK_EXIT),
```
Comment (sorentwo) [inline test/integration/executing_test.exs:74]: Invariably I mess up any suggestion I commit...

```suggestion
            action <- member_of(~w(OK DISCARD ERROR EXIT FAIL KILL SNOOZE TASK_ERROR TASK_EXIT)),
```

**Comment by derekkraan** [comment]
@sorentwo not sure if you intended to close this PR?

**Comment by sorentwo** [comment]
Nope. That was a side effect of changing the primary branch to `main`. Please re-open!

**Comment by derekkraan** [comment]
I think I have to open a new PR. I'll do that tomorrow.

**Pull request closed without merging**